### DISCUSSION TITLE
Code signing an app update for an app that already uses Play App Signing

### DISCUSSION THREAD

**Discussion by arvinio** [codemagic-ci-cd/codemagic-docs#2117] [Q&A]
### If you are looking for help, please confirm the following...

- [X] I confirm I have searched the [Docs](https://docs.codemagic.io), [Codemagic Sample Projects](https://github.com/codemagic-ci-cd/codemagic-sample-projects), and GitHub [Discussions](https://github.com/orgs/codemagic-ci-cd/discussions).
- [X] I confirm the software versions on my local computer match the ones in CI environment (e.g. Xcode version, Gradle version). See [here](https://docs.codemagic.io/troubleshooting/common-android-issues/#builds-work-locally-but-fail-on-codemagic)
- [X] I confirm I am able to do a fresh clone of the repository and build the project on my local computer.

### Which mobile framework are you using?

Flutter (Dart)

### Steps to reproduce

- Took over this project with no access to past developers
- Reset the Upload key certificate in Google Play Console
- Added the upload keystore file to CodeMagic. Both in Teams>MyTeam>Code signing identities and in the build settings of the app (what is the difference here?)
- Built app
- Uploading .aab file manually to Google Play Console
- Get the error below. Note that the expected fingerprint matches my upload keystore.
- In my process of troubleshooting this I've also requested an upgrade to App signing key certificate, and now I've uploaded both .jks files to my Team's Code Signing Identities in CodeMagic. Still now sure what this does if the same settings are duplicated in the apps Build Settings.

Error:
Your Android App Bundle is signed with the wrong key. Ensure that your App Bundle is signed with the correct signing key and try again. Your App Bundle is expected to be signed with the certificate with fingerprint:
SHA1: DA***
but the certificate used to sign the App Bundle you uploaded has fingerprint:
SHA1: 18***

### Expected results

Build without error

### Actual results

Error

Your Android App Bundle is signed with the wrong key. Ensure that your App Bundle is signed with the correct signing key and try again. Your App Bundle is expected to be signed with the certificate with fingerprint:
SHA1: DA***
but the certificate used to sign the App Bundle you uploaded has fingerprint:
SHA1: 18***

**Comment by joeboylson** [comment]
I am also having this same issue. Here is what I have tried so far:

1. generated a new keystore locally
2. In the **App Signing** section in the Google Play Console, I have uploaded this new keystore in PEM format in the "Request upload key reset". After a few days, I saw that the correct key was showing in Google Play Console _and_ my error messages are now expecting that new key.
3. I have uploaded this same keystore in the "Android code signing" in Codemagic and I have confirmed that the alias and password are correct.

Result:
- I am still seeing the same "signed with wrong key" message
- After deleting the workflow and recreating it (thought, maybe a caching issue?), I am _still_ seeing this error message - however, it now seems to be a different SHA1 key its using. Previous it was `39***`, now it is `98***`. The correct key should begin with `18***`. This test with 2 different workflows and 2 different keys leads me to believe there may be something wrong with the Android signing step.
- I have tried this with manual (downloading from codemagic and manually creating release) and automatic deployment.

@arvinio Please let me know if you find a solution - I will do likewise.

---

EDIT:

I have seen some people use this method of designating the release key in `~/.gradle/gradle.properties`. Is this required for Codemagic to use the correct key? Ideally I would like to keep the key out of my repository.

```
RELEASE_STORE_FILE=path/to/release.key
RELEASE_STORE_PASS=******
RELEASE_ALIAS=******
RELEASE_KEY_PASS=******
```

**Comment by joeboylson** [comment]
@arvinio 

I have found the solution!

I found [this guide](https://docs.codemagic.io/yaml-quick-start/migrating-from-bitrise/) which gave me some information on how to set up the proper environment variables and use the `codemagic.yaml` method instead of the workflow method. Its a little weird that it was in this "Migrating from Bitrise" article (which does not apply for me), but after following the steps, it seemed to work great!

I would recommend you try this out! 

Would be happy to provide more information if you run into any troubles.

**Reply by Souffleur-de-Reves** [reply]
Hi @arvinio 

More details would be appreciated. Did you put that part of code 

`...
  android {
      ...
      defaultConfig { ... }
      signingConfigs {
          release {
              if (System.getenv()["CI"]) { // CI=true is exported by Codemagic
                  storeFile file(System.getenv()["CM_KEYSTORE_PATH"])
                  storePassword System.getenv()["CM_KEYSTORE_PASSWORD"]
                  keyAlias System.getenv()["CM_KEY_ALIAS"]
                  keyPassword System.getenv()["CM_KEY_PASSWORD"]
              } else {
                  keyAlias keystoreProperties['keyAlias']
                  keyPassword keystoreProperties['keyPassword']
                  storeFile keystoreProperties['storeFile'] ? file(keystoreProperties['storeFile']) : null
                  storePassword keystoreProperties['storePassword']
              }
          }
      }
      buildTypes {
          release {
              ...
              signingConfig signingConfigs.release
          }
      }
  }
  ...`

And did you base64 encode your keystore file ?

**Comment by arvinio** [comment]
> @arvinio
> 
> I have found the solution!
> 
> I found [this guide](https://docs.codemagic.io/yaml-quick-start/migrating-from-bitrise/) which gave me some information on how to set up the proper environment variables and use the `codemagic.yaml` method instead of the workflow method. Its a little weird that it was in this "Migrating from Bitrise" article (which does not apply for me), but after following the steps, it seemed to work great!
> 
> I would recommend you try this out!
> 
> Would be happy to provide more information if you run into any troubles.

@joeboylson could you please share your own `codemagic.yaml` here?

**Reply by dtrdic** [reply]
hey @arvinio, do you still having same issue?
Can you please doublcheck if your signing configuration in `build.gradle` looks like [this in the docs](https://docs.codemagic.io/yaml-code-signing/signing-android/#signing-android-apps-using-gradle)?

**Comment by dtrdic** [comment]
I'm tentatively closing the discussion due to a lack of activity. You can open the discussion again by replying to this message or opening a new discussion.
